### PR TITLE
Client API update

### PR DESCRIPTION
**Pull request by dmitrykuzmin**
This PR updates the API of Spine web client.

Now, the API is a bit more flexible and tries to mimic the new Java [API](https://github.com/SpineEventEngine/core-java/pull/1186) where appropriate.

In general, the new API looks as follows:
```javascript
// Query.
client.select(Customer.class)
      .byId(westCoastCustomerIds())
      .withMask("name", "address", "email")
      .where([Filters.eq("type", "permanent"),
              Filters.eq("discount_percent", 10),
              Filters.eq("company_size", Company.Size.SMALL)])
      .orderBy("name", OrderBy.Direction.ASCENDING)
      .limit(20)
      .run(); // The returned type is `Promise<Customer[]>`.

// Subscription to entities.
client.subscribeTo(Task.class)
      .where(Filters.eq("status", Task.Status.ACTIVE))
      // Additional filtering can be done here.
      .post()
      .then(({itemAdded, itemChanged, itemRemoved, unsubscribe}) => {
          itemAdded.subscribe(_addDisplayedTask);
          itemChanged.subscribe(_changeDisplayedTask);
          itemRemoved.subscribe(_removeDisplayedTask);
      });

// Subscription to events.
client.subscribeToEvent(TaskCreated.class)
      .where([Filters.eq("task_priority", Task.Priority.HIGH),
              Filters.eq("context.past_message.actor_context.actor", userId)])
      .post()
      .then(({eventEmitted, unsubscribe}) => {
          eventEmitted.subscribe(_logEvent);
      });

// Command.
client.command(logInUser)
      .onOk(_logOk)
      .onError(_logError)
      .observe(UserLoggedIn.class, ({subscribe, unsubscribe}) => {
          subscribe(event => _logAndUnsubscribe(event, unsubscribe));
          setTimeout(unsubscribe, EVENT_WAIT_TIMEOUT);
      })
      .observe(UserAlreadyLoggedIn.class, (({subscribe, unsubscribe}) => {
          subscribe(event => _warnAboutAndUnsubscribe(event, unsubscribe));
          setTimeout(unsubscribe, EVENT_WAIT_TIMEOUT);
      })
      .post();
```
### Notable changes

* The client now allows to observe the events that are the immediate result of handling the posted command. This can be done using the `observe` method, in a similar way to the example above. Note, that in the example we make sure the event subscriptions are always cancelled, even if the event of the corresponding type is never received. This can be done in some other way but all the subscriptions created with `observe` should be cleaned up manually.

* The event subscriptions will now always write `spine.core.Event` instances to the Firebase database and always receive `spine.core.Event` instances instead of event messages on the client, in the subscription objects.

* A service which keeps the active subscriptions "alive", will no longer ignore the erroneous response received from the server. If the subscriptions can't be "kept-up", it will be removed from the list of active subscriptions and the "keep-up" requests will no longer be sent for it.

* The entity subscription will now be notified about the no-longer-matching objects with the help of `itemRemoved` callback. Previously, it was done with the `itemChanged` callback with `null` body, which raised some problems when it was necessary to identify the object which stopped matching the subscription.

#### Removed
* `Client.fetch(...)` - please use `Client.select(...).byId(...).run()` instead;
* `Client.subscribe(...)` - please use `Client.subscribeTo[Event](...).byId(...).post()` instead.

#### Renamed
* `Client.subscribeTo(Topic)` -> `Client.subscribe(Topic)`.

#### Deprecated
* `Client.execute(Query)` - please use `Client.read(Query)` instead;
* `Client.sendCommand(...)` - please use `Client.command(...).onOk(...) ... .post()` instead.

All other API remains intact.

### Other

Spine version advances to `1.2.9`.

### CODE REVIEW

**Comment by dmitrykuzmin** [comment]
@dmdashenkov PTAL. 

The coverage is low because the changes are mostly covered by the integration tests which do not count towards coverage.

File: client-js/main/client/client-request.js [hunk 1]
```diff
@@ -0,0 +1,587 @@
...
+ * A request from a client to the Spine backend.
+ *
+ * @abstract
+ */
+class ClientRequest {
+
+    /**
+     * @protected
+     */
+    constructor(client, requestFactory) {
```
Comment (dmdashenkov) [inline client-js/main/client/client-request.js]: Please fix the indent.

File: client-js/test/client/filtering-request-test.js [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
```
Comment (dmdashenkov) [inline client-js/test/client/filtering-request-test.js]: Please remove this line.

File: firebase-web/src/main/java/io/spine/web/firebase/subscription/UpdatePayload.java [hunk 3]
```diff
@@ -136,7 +140,30 @@ private static String key(Message id) {
...
+            node.addNullChild(id);
+        } else {
+            node.addChild(id, StoredJson.encode(message));
+        }
+    }
+
+    /**
+     * An identity {@link Event}-to-{@link Message} conversion.
+     */
+    private static Function<Event, Message> identity() {
```
Comment (dmdashenkov) [inline firebase-web/src/main/java/io/spine/web/firebase/subscription/UpdatePayload.java]: Please use `java.util.Function.identity()` instead.

File: integration-tests/js-tests/test/firebase-client/send-command-test.js [hunk 4]
```diff
@@ -21,14 +21,18 @@
...
 import {fail} from '../test-helpers';
 import {client, initClient} from './given/firebase-client';
+import {AnyPacker} from '@lib/client/any-packer';
+import {Type} from '@lib/client/typed-message';
 
 describe('FirebaseClient command sending', function () {
 
     // Big timeout allows to receive model state changes during tests.
-    this.timeout(5000);
+    this.timeout(15000);
```
Comment (dmdashenkov) [inline integration-tests/js-tests/test/firebase-client/send-command-test.js]: Can this be reverted? Is there a performance issue?

File: integration-tests/js-tests/test/firebase-client/send-command-test.js [hunk 5]
```diff
@@ -66,45 +76,124 @@ describe('FirebaseClient command sending', function () {
...
-                    fail(done, e.message)
-                }
-            },
-            fail(done, 'A command was rejected when an error was expected.'));
+        const checkError = error => {
+            try {
+                assert.ok(error instanceof CommandValidationError);
+                assert.ok(error.validationError());
+                // TODO:2019-06-05:yegor.udovchenko: Find the reason of failing assertion
+                // assert.ok(error.assuresCommandNeglected());
```
Comment (dmdashenkov) [inline integration-tests/js-tests/test/firebase-client/send-command-test.js]: Please check if this still fails.

**Comment by dmitrykuzmin** [comment]
@dmdashenkov PTAL again.

File: client-js/main/client/client-request.js [hunk 6]
```diff
@@ -0,0 +1,600 @@
...
+ * @property {parameterlessCallback} unsubscribe the callback which allows to cancel the
+ *                                               subscription
+ */
+
+/**
+ * A request from a client to the Spine backend.
+ *
+ * @abstract
+ */
+class ClientRequest {
```
Comment (dmdashenkov) [inline client-js/main/client/client-request.js]: Could you split this file up? It's LoC count is at 600 already. GitHub even considers it too large to display.

File: client-js/main/client/client.js [hunk 7]
```diff
@@ -136,13 +229,16 @@ export class Client {
    *
    * @example
    * // Build a subscription topic for `UserTasks` domain entity, selecting the instances
-   * // which task count is greater than 3.
+   * // whose task count is greater than 3.
```
Comment (dmdashenkov) [inline client-js/main/client/client.js]:
```suggestion
   * // with over 3 tasks.
```

File: client-js/main/client/composite-client.js [hunk 8]
```diff
@@ -34,61 +42,100 @@ import {Client} from "./client";
...
-    }
-
-    fetch({entity: cls, byIds: ids}) {
-        return this._querying.fetch(cls, ids);
-    }
-
-    newTopic() {
-        return this._subscribing.newTopic();
-    }
+  constructor(querying, subscribing, commanding) {
```
Comment (dmdashenkov) [inline client-js/main/client/composite-client.js]: Same here, please split up this file into 2 or 3. I'd say that one class per file would be an adequate strategy.

File: firebase-web/src/main/java/io/spine/web/firebase/subscription/UpdatePayload.java [hunk 9]
```diff
@@ -136,7 +140,33 @@ private static String key(Message id) {
      */
     NodeValue asNodeValue() {
         NodeValue node = NodeValue.empty();
-        messages.forEach((id, message) -> node.addChild(id, StoredJson.encodeOrNull(message)));
+        messages.forEach((id, message) -> addChildOrNull(node, id, message));
         return node;
     }
+
+    /**
+     * Adds an {@code <ID:message>} key-value pair as a child of the given node.
```
Comment (dmdashenkov) [inline firebase-web/src/main/java/io/spine/web/firebase/subscription/UpdatePayload.java]:
```suggestion
     * Adds an {@code ID -> message} key-value pair as a child of the given node.
```

File: firebase-web/src/main/java/io/spine/web/firebase/subscription/UpdatePayload.java (hunk 3 above)
Comment (dmdashenkov) [reply firebase-web/src/main/java/io/spine/web/firebase/subscription/UpdatePayload.java]: @dmitrykuzmin, please revisit this comment.

Comment (dmitrykuzmin) [reply firebase-web/src/main/java/io/spine/web/firebase/subscription/UpdatePayload.java]: As discussed vocally, I've added a clarifying comment in the code about why we don't use the standard `Function.identity()`.

**Comment by dmitrykuzmin** [comment]
@dmdashenkov PTAL again.

File: firebase-web/src/main/java/io/spine/web/firebase/subscription/UpdatePayload.java [hunk 10]
```diff
@@ -83,23 +87,23 @@ private static UpdatePayload entityUpdates(SubscriptionUpdate update) {
...
+               : Empty.getDefaultInstance();
     }
 
     private static UpdatePayload eventUpdates(SubscriptionUpdate update) {
         ImmutableMap<String, Message> messages = update
                 .getEventUpdates()
                 .getEventList()
                 .stream()
-                .collect(toHashTable(Event::id, Event::enclosedMessage));
+                .collect(toHashTable(Event::id, identity()));
```
Comment (dmdashenkov) [inline firebase-web/src/main/java/io/spine/web/firebase/subscription/UpdatePayload.java]: Please move this to `toHashTable`.

Comment (dmitrykuzmin) [reply firebase-web/src/main/java/io/spine/web/firebase/subscription/UpdatePayload.java]: Resolved vocally.

**Pull request merged**